Minutes 2022-05-11 (#301)
solid/authorization-panel: meetings/2022-05-11.md

* If this is your first time, welcome! please introduce yourself.

### Scribes

* Matthieu Bosquet
* Elf Pavlik

### Introductions

* Laurens Debackere: Working on research with Solid Lab and digital Flanders/Flemish gvt to implement use cases (UMA using access grants and data grants from the interop spec to generate access tokens). Would like to understand how access modes (acl:append and acl:update) really mean from an authorization service's point of view. Did mostly WAC work rather than ACP.

---

## Topics

### PR 300: Deprecation of ACP access

* See also: [PR 300](https://github.com/solid/authorization-panel/pull/300) and [issue 277](https://github.com/solid/authorization-panel/issues/277).

* MB: It is basically getting rid of a redundant property and aligning on WAC.
* JB: It did always feel like a piece that wouldn't really be needed. Good for me.
* EP: I just approved it. In the issue, my main concern was about access for the access control, but it didn't seem like a common requirement.

### Access Modes

https://github.com/laurensdeb/interoperability/issues/42
https://github.com/solid/authorization-panel/issues/253

> * LD: The implementation I've done is based on the CSS which uses in terms of access modes uses Read Write Append and there is the new Update mode used in the interop which isn't used anywhere else in Solid. I'm not sure what the difference between Update and Append.

Reviewer: ```suggestion
* LD: The implementation I've done is based on the CSS which uses access modes `Read` / `Write` / `Append`, and there is the new `Update` mode used in interop which isn't used anywhere else in Solid. I'm not sure of the difference between `Update` and `Append`.
```

> * JB: In the interop spec, in the [access section](https://solid.github.io/data-interoperability-panel/specification/#access-scopes) there is a table enumerating each mode (Create Update and Delete). I've been pushing to get more granular modes in line with CRUD at the protocol level. I think that the current modes are a security problem. So we define what's intended by the grant.

Reviewer: ```suggestion
* JB: In the interop spec, in the [access section](https://solid.github.io/data-interoperability-panel/specification/#access-scopes), there is a table enumerating each mode (Create, Update, and Delete). I've been pushing to get more granular modes in line with CRUD at the protocol level. I think that the current modes are a security problem. So we define what's intended by the grant.
```

* ...: If we allow Data Grant based access to resources. We would need to say that access to given resource would be for example at most read. It would require to agree who is the ultimate source of the authorization.
* ...: There is also a case if data grants grants you an access over multiple resources.
* EP: I just have a quick comment that while talking about Access Modes added in access token, we don't want to get an access token for every request. We shouldn't require the client to get a new token for every request. I don't think the access token should be specific for each request.
* MB: Yes, as far as I understand, the ID token defined in Solid OIDC, you only claim the identity of requesting agent and issuer/client... But the authorization evaluation needs to happen on every request.
* EP: The request would include access token, so the authz server can correlate what it knows about the user.
* LD: I think it's quite hard to prevent token re-issuance. I'll get back to my implementation [on GitHub](https://github.com/laurensdeb/interoperability). Right now, you need token reissuance with every request.

> * ...: When you get a 401 with a token previously got, you can add claims to your token and get new tokens with lesser overhead but it's quite complex to avoid reissuance. You get a ticket with the access modes requested. The access tokens pass information between AS and RS and also give to the token a JTI (token id). You could speed up the work of the UMA AS.

Reviewer: ```suggestion
* ...: When you get a 401 with a token previously used, you can add claims to your token and get new tokens with less overhead, but it's quite complex to avoid reissuance. You get a ticket with the access modes requested. The access tokens pass information between AS and RS and also give to the token a JTI (token id). You could speed up the work of the UMA AS.
```

* EP: There is an incoming request and then you use a backchannel.

> * LD: Indeed you could do that, but the main issue you have right now is that the CSS has to evaluate which access modes are granted over a specific resource. Which makes token re-issuance a necessity. Unless you defined an opaque token with a token introspection endpoint.
> * EP: That's my kind of point. I think the token should be issued based on the claims pushed. With requests being made, the RS says which modes are allowed for which resources.

Reviewer: ```suggestion
* LD: Indeed you could do that, but the main issue right now is that the CSS has to evaluate which access modes are granted over a specific resource, which makes token re-issuance a necessity, unless you defined an opaque token with a token introspection endpoint.
* EP: That's kind of my point. I think the token should be issued based on the claims pushed. With requests being made, the RS says which modes are allowed for which resources.
```

> * LD: You could have a session id in UMA specific to a RS and via introspection optimise how modes are granted. When using UMA, you should ideally not have to trigger issuance for every request. But I'm still unsure about how to implement it.

Reviewer: ```suggestion
* LD: You could have a session id in UMA specific to an RS, and optimise how modes are granted via introspection. When using UMA, you should ideally not have to trigger issuance for every request. But I'm still unsure about how to implement it.
```

> * EP: With UMA, different resources could be protected by different AS. I think with UMA, you could in theory in 401 for different resource return differen AS usign `as_uri`

Reviewer: ```suggestion
* EP: With UMA, different resources could be protected by different AS. I think with UMA, you could in theory for different resource return different AS usign `as_uri` in 401
```

* LD: We want the AS to perform authorization. And it's not specified anywhere.

> * MB: Both WAC and ACP have a very resource centric approach, so you evaluate access based on which resource you try to access rather than which resource a user has access to.

Reviewer: ```suggestion
* MB: Both WAC and ACP have a very resource-centric approach, so you evaluate access based on which resource you try to access, rather than which resource a user has access to.
```

> * JB:
> * LD: The hardest part with respecting the UMA spec is it doesn't really specify what happens when you get outside of the boundary of your token (token upgrade), it only says get a 403 (not a www header with an UMA ticket). That's mainly why every request comes with its own flow. Otherwise, there is no way of figuring out what the limits of your ticket really are. I think there is security implications for the choice of a 403 rather than 401 in UMA.

Reviewer: ```suggestion
* LD: The hardest part with respect to the UMA spec is it doesn't really specify what happens when you get outside of the boundary of your token (token upgrade); it only says get a 403 (not a www header with an UMA ticket). That's mainly why every request comes with its own flow. Otherwise, there is no way of figuring out what the limits of your ticket really are. I think there are security implications for the choice of a 403 rather than 401 in UMA.
```

* MB: It's complicated.
* LD: I'm trying to get UMA modules into CSS.

> * EP: I don't think we were aiming for full UMA compliance. We mostly wanted the claim pushing part. From RS we pretty much want the AS and IRI. I don't think we should expect that RS is fully UMA compliant.

Reviewer: ```suggestion
* EP: I don't think we were aiming for full UMA compliance. We mostly wanted the claim pushing part. From RS, we pretty much want the AS and IRI. I don't think we should expect that RS is fully UMA compliant.
```

> * MB: In terms of what we are trying to achieve here. We shouldn't tightly couple ACP or WAC or Data Grants to UMA. We can agree that it's good to have access modes interface that needs to be standard across all authorization systems. Access modes entail operations.

Reviewer: ```suggestion
* MB: In terms of what we are trying to achieve here, we shouldn't tightly couple ACP or WAC or Data Grants to UMA. We can agree that it's good to have an access modes interface that needs to be standard across all authorization systems. Access modes entail operations.
```
